paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, 
+users can build a real network from a `NetDef` in protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetworkBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class NetworkBase {
+ public:
+  // `def` is a proto message that describe the structure of a network.
+  NetworkBase(const NetDef& def);
```

Comment:
Maybe `addOperator` method is very useful, especially when we generating backward ops and generating optimization ops.

Also, the constructor of `NetworkBase` should invoke `addOperator` as well.

```cpp
class NetworkBase {
 public:
  Error AddBackwardOps(const std::string& lossVariableName, float lossGrad = 1.0f) {
    for (eachOp reversely) {
      for (gradOp in GetGradientOps(eachOp)) {
        addOperator(gradOp)
      }
    }
  }

  Error AddOptimizerOps(const Map<string, string> paramToGradMap) {
    for (paramName, gradName in paramToGradMap) {
      addOperator(SGD(input={gradName, paramName}, output={paramName}));
    }
  }
};

```